### PR TITLE
Add function to convert Datetime from a dictionary to Epoch

### PR DESCRIPTION
**Pull request by Razzlegames**
- Also changed get_time_from_unix_time to get_datetime_from_unix_time to be
  consistent.

Ticket:
https://github.com/godotengine/godot/issues/4038

### CODE REVIEW

**Comment by Razzlegames** [comment]
Note that this breaks ``get_time_from unix_time`` and will now be known as ``get_date_time_from_unix_time``

**Comment by Razzlegames** [comment]
## Test using

```

extends Spatial

func _ready():
	
	set_process(true)
	print("Time: "+ str(OS.get_time()))
	print("Date: "+ str(OS.get_date()))
	
	pass

func testReverseConversion():
	
	var time = OS.get_unix_time()
	var dict = OS.get_date_time_from_unix_time(time)
	var testTime = OS.get_unix_time_from_date_time(dict)
	print("testTime: "+ str(testTime) + \
	  " Actual time: "+ str(time))
	
	if(testTime == time):
		print("Epoch reverse calculation was CORRECT!");
	else:
		print("!!!!!! EPOCH CALC WRONG !!!!!!");
		
	# Test error case
	dict.clear()
	var wrongTime = OS.get_unix_time_from_date_time(dict)
	
	print(str(wrongTime))
		
	pass

func _process(delta):
	
	testReverseConversion()
	var time = OS.get_unix_time()
	var dict = OS.get_date_time_from_unix_time(time)
	
	var dateDict = OS.get_date()
	if(OS.MONTH_MARCH != dateDict['month']):
		print("month: "+ str(dateDict['month']) + \
		   " but OS.MONTH_MARCH = "+ str(OS.MONTH_MARCH))
	
	print(dict) 
```

## Output

```
** Debug Process Started **
Time: (hour:21), (minute:38), (second:29)
Date: (day:15), (dst:False), (month:3), (weekday:2), (year:2016)
testTime: 1458103109 Actual time: 1458103109
Epoch reverse calculation was CORRECT!
-86400
```

**Comment by akien-mga** [comment]
How about using ``get_date_from_unix_time`` and ``get_unix_time_from_date`` as names? ``date_time`` is a bit redundant, and the OS functions are already named ``get_date`` and ``get_unix_time``, so that fits.

**Comment by Razzlegames** [comment]
It is a bit wordy but the reason is, both get_date and get_time functions return less than these functions. Date time functions are really a combination of both (but missing DST). This the reason for the name.

**Comment by akien-mga** [comment]
But I mean the existing functions are using ``date`` and ``unix_time``, so when combining the two it should be ``get_date_from_unix_time`` and not ``get_date_**time_**from_unix_time``.

**Comment by Razzlegames** [comment]
But doesn't that imply you could just pass in the results of getting date, and receive accurate epoch?  That's not the case here.  I'm for a better more descriptive name, but I think picking just date is confusing.

**Comment by akien-mga** [comment]
So IIUC the return of ``get_date()`` and ``get_date_time_from_unix_time(get_unix_time())`` would not be the same? If so that's quite confusing already IMO even if it uses the name ``date_time`` to be differenciated from ``get_date``.

I might have to recheck your previous PR to see what was the rationale for adding a different date format, but ideally there should be only two formats (``get_date()`` and ``get_unix_time()``), and helper functions to go from one format to the other (``get_date_from_unix_time()`` and ``get_unix_time_from_date()``).

**Comment by Razzlegames** [comment]
> So IIUC the output of get_date() and get_date_time_from_unix_time(get_unix_time()) would not be the same?

That's right.  That's why the name change.  ``get_date`` returns far less information (no seconds, hours, minutes).  I'm following the convention of many languages here to use the name date time to have the specific meaning of granular time and calendar date.  

Python comes to mind: https://docs.python.org/2/library/datetime.html

I could condense instead of ``date_time`` to ``datetime`` to be more inline.

**Comment by akien-mga** [comment]
Ok I understand. It still seems a bit convoluted to me, basically if I want the datetime I'd have to ``get_date_time_from_unix_time(get_unix_time())`` because ``get_date()`` only gives the calendar date? Wouldn't it be better to add the granular time to the dict returned by ``get_date()``?

**Comment by Razzlegames** [comment]
Possibly. But I wanted my code changes to be constrained to the problem they are intending to solve, with lowest impact. I didn't make any assumptions about the existing ``get_date`` requirements.  It could also be desirable for the user to only want calendar date, I would assume. 

Also, I have a feeling (need to look again) that all get_date functions are impkented on an OS by OS basis. I didn't want to go mucking around all the different platforms, since I don't have a way to test them all (even though it would hopefully just involve deleting code in those areas, in this case)

**Comment by Razzlegames** [comment]
I've got it. I'll create a ``get_datetime`` function and change all naming to ``datetime``, a clearer and distinct type.  Now there will be functions to convert ``datetime`` dictionaries to ``unix_time`` and back :)

**Comment by akien-mga** [comment]
Sounds much clearer to me, thanks! :)

**Comment by blurymind** [comment]
So If I make my own dictionary with a date that is not today and now, this new function will be able to convert it to unix time stamp. Then  I will be able to apply any maths (involving other unix timestamps) on it and then convert the result back to a human readable date time dictionary :)

Can't wait for this to get in trunk! Thank you for implementing it.

**Comment by Razzlegames** [comment]
@blurymind  Yep, you can convert to unix time stamp given a dictionary of:  year, month, day, hour, minute, second

**Comment by Razzlegames** [comment]
@akien-mga I think I've covered all the corner cases and things.  Please let me know if you see anything.  I also added some error handling to be a bit more user friendly. :)

**Comment by blurymind** [comment]
you guys are awesome!  
@Razzlegames  you aced it. Compiling now to get it in my todo app. :)
